java/rocksjni/rocksjni.cc

```diff
@@ -746,6 +747,282 @@ void Java_org_rocksdb_RocksDB_putDirect(
+    const ROCKSDB_NAMESPACE::WriteOptions default_write_options =
+        ROCKSDB_NAMESPACE::WriteOptions();
+
+    auto columns_family =
+        (jcf_handle == 0)
+            ? db->DefaultColumnFamily()
+            : reinterpret_cast<ROCKSDB_NAMESPACE::ColumnFamilyHandle*>(
+                  jcf_handle);
+
+    db->PutEntity(default_write_options, columns_family, key, columns);
+  };
+
+  ROCKSDB_NAMESPACE::WideColumnJni::convertWideColumns(
+      putEntity, env, jkey, keyOffset, keyLen, jnames, jvalues);
+}
+
+/*
+ * Class:     org_rocksdb_RocksDB
+ * Method:    putEntity
+ * Signature:
+ * (JLjava/nio/ByteBuffer;II[Ljava/nio/ByteBuffer;[I[I[Ljava/nio/ByteBuffer;[I[IJ)V
+ */
+void Java_org_rocksdb_RocksDB_putEntityDirect(
+    JNIEnv* env, jclass, jlong jdb_handle, jobject jKey, jint key_offset,
+    jint key_len, jobjectArray jnames, jintArray jnames_offset,
+    jintArray jnames_len, jobjectArray jvalues, jintArray jvalues_offset,
+    jintArray jvalues_len, jlong jcf_handle) {
+  auto* db = reinterpret_cast<ROCKSDB_NAMESPACE::DB*>(jdb_handle);
+
+  auto putEntity = [&db, &jcf_handle](ROCKSDB_NAMESPACE::Slice key,
+                                      ROCKSDB_NAMESPACE::WideColumns columns) {
+    const ROCKSDB_NAMESPACE::WriteOptions default_write_options =
+        ROCKSDB_NAMESPACE::WriteOptions();
+
+    auto columns_family =
+        (jcf_handle == 0)
+            ? db->DefaultColumnFamily()
+            : reinterpret_cast<ROCKSDB_NAMESPACE::ColumnFamilyHandle*>(
+                  jcf_handle);
+
+    db->PutEntity(default_write_options, columns_family, key, columns);
+  };
+
+  ROCKSDB_NAMESPACE::WideColumnJni::convertWideColumnsDirect(
+      putEntity, env, jKey, key_offset, key_len, jnames, jnames_offset,
+      jnames_len, jvalues, jvalues_offset, jvalues_len);
+}
+
+/*
+ * Class:     org_rocksdb_RocksDB
+ * Method:    getEntityDirect
+ * Signature:
+ * (JLjava/nio/ByteBuffer;II[Ljava/nio/ByteBuffer;[I[I[I[Ljava/nio/ByteBuffer;[I[I[I)Lorg/rocksdb/Status;
+ */
+jobject Java_org_rocksdb_RocksDB_getEntityDirect(
+    JNIEnv* env, jobject, jlong jdb_handle, jobject jKey, jint key_offset,
+    jint key_len, jobjectArray jnames, jintArray jNamesOffset,
+    jintArray jNamesLen, jintArray jnamesRequiredSize, jobjectArray jvalues,
+    jintArray jValuesOffset, jintArray jValuesLen,
+    jintArray jvaluesRequiredSize, jlong jcf_handle) {
+  auto* db = reinterpret_cast<ROCKSDB_NAMESPACE::DB*>(jdb_handle);
+  auto columns_family =
+      (jcf_handle == 0)
+          ? db->DefaultColumnFamily()
+          : reinterpret_cast<ROCKSDB_NAMESPACE::ColumnFamilyHandle*>(
+                jcf_handle);
+
+  auto _key = env->GetDirectBufferAddress(jKey);
+  if (_key == nullptr) {
+    ROCKSDB_NAMESPACE::RocksDBExceptionJni::ThrowNew(
+        env,
+        "Invalid key argument (argument is not a valid direct ByteBuffer)");
+    return nullptr;
+  }
+  auto key = reinterpret_cast<char*>(_key) + key_offset;
```

Comment:
Could this use portal.h for string conversion

Reply:
No. See previous comment